subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -237,40 +260,65 @@ public FileCollection fileCollection(Dependency... dependencies) {
         return new ConfigurationFileCollection(WrapUtil.toLinkedSet(dependencies));
     }
 
-    public void includedInResolveResult() {
-        includedInResult = true;
+    public void markAsObserved() {
+        synchronized (lock) {
+            if (state == InternalState.UNOBSERVED) {
+                state = InternalState.OBSERVED;
+            }
+        }
         for (Configuration configuration : extendsFrom) {
-            ((ConfigurationInternal) configuration).includedInResolveResult();
+            ((ConfigurationInternal) configuration).markAsObserved();
         }
     }
 
     public ResolvedConfiguration getResolvedConfiguration() {
-        resolveNow();
+        resolveNow(InternalState.RESULTS_RESOLVED);
         return cachedResolverResults.getResolvedConfiguration();
     }
 
-    private void resolveNow() {
+    private void resolveNow(InternalState requestedState) {
         synchronized (lock) {
-            if (state == State.UNRESOLVED) {
+            boolean needsResolve = state == InternalState.UNOBSERVED || state == InternalState.OBSERVED || modified;
```

Comment:
How about this?

```java
boolean needsResolve = false;
switch (state) {
    case UNOBSERVED:
    case OBSERVED:
        needsResolve = true;
        break;
    case TASK_DEPENDENCIES_RESOLVED:
        needsResolve = modified && requestedState == InternalState.RESULTS_RESOLVED;
        if (modified && !needsResolve) {
            DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to resolve already resolved configuration %s. Changes made since the configuration was originally resolved are ignored", getDisplayName()));
        }
        break;
    case RESULTS_RESOLVED:
        if (modified) {
            DeprecationLogger.nagUserOfDeprecatedBehaviour(String.format("Attempting to resolve already resolved configuration %s. Changes made since the configuration was originally resolved are ignored", getDisplayName()));
        }
        break;
}
if (needsResolve) {
   // ...
}
```

This would also take care of nagging the user about the changes after resolution being ignored. *And* it would give the user the location where the re-resolve was started from, which would help in debugging.